cli/internal/lockfile/yarn_lockfile.go

```diff
@@ -73,93 +51,39 @@ func (l *YarnLockfile) AllDependencies(key string) (map[string]string, bool) {
 
 // Subgraph Given a list of lockfile keys returns a Lockfile based off the original one that only contains the packages given
 func (l *YarnLockfile) Subgraph(packages []string) (Lockfile, error) {
-	lockfile := make(YarnLockfile, len(packages))
+	lockfile := make(map[string]yarnlock.LockFileEntry, len(packages))
 	for _, key := range packages {
-		entry, ok := (*l)[key]
+		entry, ok := (l.inner)[key]
 		if ok {
 			lockfile[key] = entry
 		}
 	}
 
-	return &lockfile, nil
+	return &YarnLockfile{lockfile, l.hasCLRF}, nil
 }
 
 // Encode encode the lockfile representation and write it to the given writer
 func (l *YarnLockfile) Encode(w io.Writer) error {
-	var tmp bytes.Buffer
-	yamlEncoder := yaml.NewEncoder(&tmp)
-	yamlEncoder.SetIndent(2)
-	if err := yamlEncoder.Encode(l); err != nil {
-		return errors.Wrap(err, "failed to materialize sub-lockfile. This can happen if your lockfile contains merge conflicts or is somehow corrupted. Please report this if it occurs")
+	writer := w
+	if l.hasCLRF {
+		writer = crlf.NewWriter(w)
 	}
-
-	if _, err := io.WriteString(w, "# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n# yarn lockfile v1\n\n"); err != nil {
-		return errors.Wrap(err, "failed to write to buffer")
-	}
-
-	// because of yarn being yarn, we need to inject lines in between each block of YAML to make it "valid" SYML
-	scan := bufio.NewScanner(&tmp)
-	buf := make([]byte, 0, 1024*1024)
-	scan.Buffer(buf, 10*1024*1024)
-	for scan.Scan() {
-		line := scan.Text() //Writing to Stdout
-		var stringToWrite string
-		if !strings.HasPrefix(line, " ") {
-			stringToWrite = fmt.Sprintf("\n%v\n", strings.ReplaceAll(line, "'", "\""))
-		} else {
-			stringToWrite = fmt.Sprintf("%v\n", strings.ReplaceAll(line, "'", "\""))
-		}
-		if _, err := io.WriteString(w, stringToWrite); err != nil {
-			return errors.Wrap(err, "failed to write to buffer")
-		}
+	if err := l.inner.Encode(writer); err != nil {
+		return errors.Wrap(err, "Unable to encode yarn.lock")
 	}
-
 	return nil
 }
 
 // DecodeYarnLockfile Takes the contents of a yarn lockfile and returns a struct representation
 func DecodeYarnLockfile(contents []byte) (*YarnLockfile, error) {
-	var lockfile map[string]*YarnLockfileEntry
-
-	var next []byte
-	var lines []string
-	var l *regexp.Regexp
-	var output string
-
-	hasLF := !bytes.HasSuffix(contents, []byte("\r\n"))
-	if hasLF {
-		lines = strings.Split(string(contents), "\n")
-		l = nLineEnding
-	} else {
-		lines = strings.Split(strings.TrimRight(string(contents), "\r\n"), "\r\n")
-		l = rnLineEnding
-	}
+	lockfile, err := yarnlock.ParseLockFileData(contents)
+	hasCLRF := bytes.HasSuffix(contents, crlfLiteral)
```

Comment:
Is just the final suffix good enough? I feel like this could be possibly trimmed by something like editorconfig.

Reply:
What if we added this as a metadata thing from the parser? We know for sure at this line:
https://github.com/iseki0/go-yarnlock/blob/master/parser.go#L109